### PR TITLE
CLI: Add support for Nuxt to project init

### PR DESCRIPTION
**Pull request by tobiasdiez**
## What I did

Add support for nuxt projects when initializing a new project. This uses the nuxt module https://github.com/nuxt-modules/storybook and the underlying framework implementation https://github.com/storybook-vue/storybook-nuxt.

This module, as most nuxt modules, follows a "batteries included" approach and provides all necessary dependencies. In particular, `storybook` and the storybook-nuxt framework. Thus, a few changes needed to be done in the `baseGenerator` code to disable a few features for the nuxt integrtion.

#### The changes in this PR are covered in the following automated tests:
- [ ] stories
- [x] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

Create a new project using `npx nuxi@latest init` followed by `code\lib\cli\bin\index.js init`.

### Documentation

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

This pull request has been released as version `0.0.0-pr-26884-sha-1ec720b9`. Try it out in a new sandbox by running `npx storybook@0.0.0-pr-26884-sha-1ec720b9 sandbox` or in an existing project with `npx storybook@0.0.0-pr-26884-sha-1ec720b9 upgrade`.
<details>
<summary>More information</summary>

| | |
| --- | --- |
| **Published version** | [`0.0.0-pr-26884-sha-1ec720b9`](https://npmjs.com/package/@storybook/cli/v/0.0.0-pr-26884-sha-1ec720b9) |
| **Triggered by** | @kasperpeulen |
| **Repository** | [tobiasdiez/storybook](https://github.com/tobiasdiez/storybook) |
| **Branch** | [`nuxt-generator`](https://github.com/tobiasdiez/storybook/tree/nuxt-generator) |
| **Commit** | [`1ec720b9`](https://github.com/tobiasdiez/storybook/commit/1ec720b96fd13330e61db6181fe00dc844ad7b96) |
| **Datetime** | Fri May 17 08:58:09 UTC 2024 (`1715936289`) |
| **Workflow run** | [9125643910](https://github.com/storybookjs/storybook/actions/runs/9125643910) |

To request a new release of this pull request, mention the `@storybookjs/core` team.

_core team members can create a new canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=26884`_
</details>

### CODE REVIEW

**Comment by yannbf** [comment]
Hey @tobiasdiez thanks for your contribution! Seems like a duplicate of #23663. I'll let you, @kasperpeulen and @chakAs3 discuss about it. Thanks!

**Comment by kasperpeulen** [comment]
Could you review this one @chakAs3 ?

**Comment by tobiasdiez** [comment]
> Hey @tobiasdiez thanks for your contribution! Seems like a duplicate of #23663. I'll let you, @kasperpeulen and @chakAs3 discuss about it. Thanks!

Ohh, I was not aware of #23663. This would have made my life easier 🙈.

Btw, https://github.com/nuxt-modules/storybook/tree/main/examples/starter was generated using the code of this PR.

**Comment by chakAs3** [comment]
> > Hey @tobiasdiez thanks for your contribution! Seems like a duplicate of #23663. I'll let you, @kasperpeulen and @chakAs3 discuss about it. Thanks!
> 
> Ohh, I was not aware of #23663. This would have made my life easier 🙈. It looks like my PR covers the same functionality and I had all the changes of #23663 except for the removal of nuxt from the unsupported template check. The latter I just pushed.
> 
> Btw, https://github.com/nuxt-modules/storybook/tree/main/examples/starter was generated using the code of this PR.

I've put in some work on it, but since there wasn't anyone available from the team to review it at the time, I decided to create a separate CLI called 'storybook-nuxt' to expedite the process and deliver something to Nuxt developers who have been eagerly anticipating it. ✌️

**Review by chakAs3 — COMMENTED** [review]
@shilman @yannbf  , could it be a good moment to transition the namespace to @storybook/nuxt, making it an official implementation? A significant amount of effort has been invested since @tobiasdiez joined me in maintaining the module and preset.

I still advocate for keeping the framework preset @storybook/nuxt separate from the monorepo. This way, we can concentrate on Nuxt-specific features without being overly dependent on Storybook. The same approach could be applied to Next.js as well; extracting them from the monorepo might facilitate faster iterations and more focused developmen

**Comment by tobiasdiez** [comment]
@yannbf @kasperpeulen could you please review this PR? Thanks!

Then we can retire our own custom cli implementation https://github.com/storybook-vue/storybook-nuxt/tree/main/packages/storybook-nuxt-cli.

File: code/lib/cli/src/generators/NUXT/index.ts
Comment (kasperpeulen) [inline code/lib/cli/src/generators/NUXT/index.ts:14]: Can you explain why we need those options only fo the nuxt package?

File: code/lib/cli/src/generators/baseGenerator.ts
Comment (kasperpeulen) [inline code/lib/cli/src/generators/baseGenerator.ts:26]: I think this option is never set to false?

**Review by kasperpeulen — COMMENTED** [review]
Added some type fixes, I have a couple of questions, and kicked off a canary so that I can do some e2e testing.

We plan to merge this into the 8.2 alpha branch after the 8.1 release Wednesday! And patch it back to 8.1 when we have some more feedback.

File: code/lib/cli/src/generators/NUXT/index.ts
Comment (tobiasdiez) [reply code/lib/cli/src/generators/NUXT/index.ts:14]: The setup with nuxt is a bit different. In the nuxt world, you normally install a so-called module. The task of the module is to interface with whatever external library is used under the hood, as well as shipping the library (i.e. it is declared as a dependency). 

For example, the tailwind module for nuxt (https://github.com/nuxt-modules/tailwindcss) declares `tailwindcss` as a dependency (https://github.com/nuxt-modules/tailwindcss/blob/main/package.json#L56). Similarly, in the storybook module we already pull-in storybook as the dependency: https://github.com/nuxt-modules/storybook/blob/30cb7036f1bfc5bc50b11b6b7c3857bcf23886b6/package.json#L81 Thus users don't need to add the storybook packages themselves in their `package.json`.

File: code/lib/cli/src/generators/baseGenerator.ts
Comment (tobiasdiez) [reply code/lib/cli/src/generators/baseGenerator.ts:26]: Right, this is not used at the moment as we still have to do a bit of work on the side of the storybook nuxt module. The idea is that by default one has a zero-config setup for storybook with nuxt (i.e. no `.storybook` folder) and the module provides the necessary config. It would also enable us to specify the storybook config via the `nuxt.config` file. This is the common pattern with nuxt, see https://nuxt.com/docs/getting-started/configuration#external-configuration-files.

Comment (kasperpeulen) [reply code/lib/cli/src/generators/baseGenerator.ts:26]: So should I wait for that work, before merging this PR?

**Comment by chakAs3** [comment]
> Btw, https://github.com/nuxt-modules/storybook/tree/main/examples/starter was generated using the code of this PR.

this noticed that the .storybook generated files are .js  instead of .ts,  However the project is a typescript project ? there is issue with project type detection when it is Nuxt project i guess

**Comment by kasperpeulen** [comment]
@chakAs3 I think our logic is mainly checking if `typescript` is a (dev) dependency of the project.

File: code/lib/cli/src/generators/baseGenerator.ts
Comment (kasperpeulen) [reply code/lib/cli/src/generators/baseGenerator.ts:26]: cc @tobiasdiez and @chakAs3

Comment (tobiasdiez) [reply code/lib/cli/src/generators/baseGenerator.ts:26]: From my side we can already merge this PR as is and once we support the "no `.storybook`" mode in the nuxt module, I'll open an new PR setting this toggle to true for nuxt projects. Would this be okay with you?

**Comment by tobiasdiez** [comment]
> @chakAs3 I think our logic is mainly checking if `typescript` is a (dev) dependency of the project.

Right, I've now added a bit of logic to check for the existence for `tsconfig.json`. This works well for nuxt. I don't know if there are any side-effects for other types of projects, but I guess you only have a tsconfig file if you really use ts.

File: code/lib/cli/src/generators/baseGenerator.ts
Comment (kasperpeulen) [reply code/lib/cli/src/generators/baseGenerator.ts]: Sounds good!

**Comment by kasperpeulen** [comment]
@tobiasdiez I have the following error (used the pnpm option of nuxi):

```
npx storybook@0.0.0-pr-26884-sha-1ec720b9 init
╭────────────────────────────────────────────────────────────────────────────╮
│                                                                            │
│   Adding Storybook version 0.0.0-pr-26884-sha-1ec720b9 to your project..   │
│   This is a pre-release version.                                           │
│                                                                            │
╰────────────────────────────────────────────────────────────────────────────╯
 • Detecting project type. ✓
Installing dependencies...

Lockfile is up to date, resolution step is skipped
Already up to date

> nuxt-app@ postinstall /Users/kasperpeulen/code/storybook/nuxt-app
> nuxt prepare

✔ Types generated in .nuxt                                                                                                                                                                          12:01:25 PM
Done in 2.2s
 • Adding Storybook support to your "Nuxt" app
  ✔ Getting the correct version of 6 packages
  ✔ Installing Storybook dependencies
. ✓
Installing dependencies...

Lockfile is up to date, resolution step is skipped
Already up to date

> nuxt-app@ postinstall /Users/kasperpeulen/code/storybook/nuxt-app
> nuxt prepare


 ERROR  Cannot read properties of undefined (reading '__STORYBOOK__')                                                                                                                                12:01:39 PM

  at setup (node_modules/.pnpm/@nuxtjs+storybook@8.0.0_nuxt@3.11.2_prettier@3.2.5_react-dom@18.3.1_react@18.3.1_vite@5.2.11_vue@3.4.27/node_modules/@nuxtjs/storybook/dist/module.mjs:122:25)
  at normalizedModule (node_modules/.pnpm/@nuxt+kit@3.11.2/node_modules/@nuxt/kit/dist/index.mjs:2112:37)
  at async installModule (node_modules/.pnpm/@nuxt+kit@3.11.2/node_modules/@nuxt/kit/dist/index.mjs:2415:95)
  at async initNuxt (node_modules/.pnpm/nuxt@3.11.2_@opentelemetry+api@1.8.0_@unocss+reset@0.60.3_floating-vue@5.2.2_unocss@0.60.3_vite@5.2.11/node_modules/nuxt/dist/index.mjs:4233:7)
  at async loadNuxt (node_modules/.pnpm/nuxt@3.11.2_@opentelemetry+api@1.8.0_@unocss+reset@0.60.3_floating-vue@5.2.2_unocss@0.60.3_vite@5.2.11/node_modules/nuxt/dist/index.mjs:4331:5)
  at async loadNuxt (node_modules/.pnpm/@nuxt+kit@3.11.2/node_modules/@nuxt/kit/dist/index.mjs:2570:19)
  at async Object.run (node_modules/.pnpm/nuxi@3.11.1/node_modules/nuxi/dist/chunks/prepare.mjs:53:18)
  at async runCommand$1 (node_modules/.pnpm/nuxi@3.11.1/node_modules/nuxi/dist/shared/nuxi.9edf0930.mjs:1648:16)
  at async runCommand$1 (node_modules/.pnpm/nuxi@3.11.1/node_modules/nuxi/dist/shared/nuxi.9edf0930.mjs:1639:11)
  at async runMain$1 (node_modules/.pnpm/nuxi@3.11.1/node_modules/nuxi/dist/shared/nuxi.9edf0930.mjs:1777:7)



 ERROR  Cannot read properties of undefined (reading '__STORYBOOK__')                                                                                                                                12:01:39 PM

 ELIFECYCLE  Command failed with exit code 1.
An error occurred while installing dependencies.

attention => Storybook now collects completely anonymous telemetry regarding usage.
This information is used to shape Storybook's roadmap and prioritize features.
You can learn more, including how to opt-out if you'd not like to participate in this anonymous program, by visiting the following URL:
https://storybook.js.org/telemetry
```

**Comment by kasperpeulen** [comment]
@tobiasdiez @chakAs3 Or do I do something wrong?

I just did:
```
npx nuxi@latest init
npx storybook@0.0.0-pr-26884-sha-1ec720b9 init
```

**Comment by chakAs3** [comment]
> 
> ```
> npx nuxi@latest init
> npx storybook@0.0.0-pr-26884-sha-1ec720b9 init
> ```

This should work as expected, but I haven't tested this PR myself. I don't see any issues. @tobiasdiez, did you have a chance to check it?

**Comment by tobiasdiez** [comment]
This looks like the bug https://github.com/nuxt-modules/storybook/issues/538 that was already fixed in the main branch but isn't yet released (@chakAs3 feel free to prepare a new release whenever you have the time for it; I don't have any other urgent fixes planned in the near future)

**Comment by AdamovichArtur** [comment]
Any updates?

**Comment by kasperpeulen** [comment]
Will merge as soon as a new version of `@nuxtjs/storybook` is released with the fix.

**Comment by chakAs3** [comment]
> Will merge as soon as a new version of `@nuxtjs/storybook` is released with the fix.

For for the delay it is released : https://github.com/nuxt-modules/storybook/releases/tag/v8.1.1

**Comment by tobiasdiez** [comment]
@kasperpeulen did you already found the time to try it out with the newest version?

**Comment by kasperpeulen** [comment]
@tobiasdiez Just fixed the merge conflicts, let's see if we can get CI green again.

**Review by kasperpeulen — APPROVED** [review]
Thanks @tobiasdiez and @chakAs3 !

File: code/lib/cli/src/generators/baseGenerator.ts
Comment (ndelangen) [inline code/lib/cli/src/generators/baseGenerator.ts:290]: @kasperpeulen I will need to roll-back this optional installation of the `storybook` package to ensure this https://github.com/storybookjs/storybook/pull/28310 works.

I'm going to just proceed, but I'd like to evaluate with you on what to do about this.

Comment (kasperpeulen) [reply code/lib/cli/src/generators/baseGenerator.ts:290]: @tobiasdiez said that it works a bit different in the nuxt world

> The setup with nuxt is a bit different. In the nuxt world, you normally install a so-called module. The task of the module is to interface with whatever external library is used under the hood, as well as shipping the library (i.e. it is declared as a dependency).

> For example, the tailwind module for nuxt (https://github.com/nuxt-modules/tailwindcss) declares tailwindcss as a dependency (https://github.com/nuxt-modules/tailwindcss/blob/main/package.json#L56). Similarly, in the storybook module we already pull-in storybook as the dependency: https://github.com/nuxt-modules/storybook/blob/30cb7036f1bfc5bc50b11b6b7c3857bcf23886b6/package.json#L81 Thus users don't need to add the storybook packages themselves in their package.json.

Not sure how that would affect that in 8.2 `storybook` is required as a peer dep though. I guess we should test it to find out.

Comment (tobiasdiez) [reply code/lib/cli/src/generators/baseGenerator.ts:290]: This shouldn't be a problem for us. The nuxt module can pull-in whatever top-level storybook dependency is appropriate. In fact, we do this already for the `storybook` package: https://github.com/nuxt-modules/storybook/blob/d1bd15a990851bbcee53822b09acd046a657fd22/packages/nuxt-module/package.json#L50

@ndelangen what issues do you expect?

**Pull request merged**
